File: ts/src/identity.ts
import * as pk from "./rsa"
import * as pem from "./pem"

export class PrivateIdentity {

  readonly decryptingKey : pk.DecryptingKey;
  readonly signingKey : pk.SigningKey;

  constructor(decryptingKey: pk.DecryptingKey, signingKey: pk.SigningKey) {
    this.decryptingKey = decryptingKey;
    this.signingKey = signingKey;
  }

  static async generate() : Promise<PrivateIdentity> {
    const [decrypter, signer] = await Promise.all([
      pk.DecryptingKey.generate(),
      pk.SigningKey.generate(),
    ]);

    return new PrivateIdentity(decrypter, signer);
  }

  publicIdentity(): PublicIdentity {
    return new PublicIdentity(this.decryptingKey.encryptingKey(),
                              this.signingKey.verifyingKey());
  }

  async sign(msg: ArrayBuffer) : Promise<pk.SignedText> {
    return this.signingKey.sign(msg);
  }

  async toJSON(): Promise<string> {
    const [decrypter, signer] = await Promise.all([
      this.decryptingKey.toPEM(),
      this.signingKey.toPEM(),
    ]);

    const identity = {
      decrypting: decrypter,
      signing: signer,
    }

    return JSON.stringify(identity);
  }

  // end PrivateIdentity
}

export class PublicIdentity {

  readonly encryptingKey : pk.EncryptingKey
  readonly verifyingKey : pk.VerifyingKey

  constructor(encryptingKey: pk.EncryptingKey, verifyingKey: pk.VerifyingKey) {
    this.encryptingKey = encryptingKey;
    this.verifyingKey = verifyingKey;
  }

  static async fromJSON(text: string) : Promise<PublicIdentity> {
    const identity = JSON.parse(text);
    if (!identity.hasOwnProperty('encrypting') || !identity.hasOwnProperty('verifying'))
      throw new Error("Text is not a public identity JSON");

    const [encrypter, verifier] = await Promise.all([
      pk.EncryptingKey.fromPublicKeyPEM(identity.encrypting),
      pk.VerifyingKey.fromPublicKeyPEM(identity.verifying)
    ]);

    return new PublicIdentity(encrypter, verifier);
  }

  async toJSON() : Promise<string> {
    const [encrypter, verifier] = await Promise.all([
      this.encryptingKey.toPEM(),
      this.verifyingKey.toPEM(),
    ]);

    const identity = {
      encrypting: encrypter,
      verifying: verifier,
    }

    return JSON.stringify(identity);
  }

  async encrypt(data: ArrayBuffer) : Promise<pk.PKCiphertext> {
    return this.encryptingKey.encrypt(data);
  }

  // end PublicIdentity
}

File: py/hippo/identity.py
import json

from .rsa import DecryptingKey, SigningKey, CipherText

class PublicIdentity:
    def __init__(self, encrypting, verifying):
        self.encrypting_key = encrypting
        self.verifying_key = verifying

    def toJSON(self):
        public_identity = {
            "encrypting": self.encrypting_key.toPEM(),
            "verifying": self.verifying_key.toPEM(),
        }

        return json.dumps(public_identity)

class PrivateIdentity:
    def __init__(self, decrypting, signing):
        self.decrypting_key = decrypting
        self.signing_key = signing

    @classmethod
    def fromJSON(cls, text):
        if isinstance(text, str):
            text = json.loads(text)
        elif not isinstance(text, dict):
            raise Exception("Expected JSON string or dict")

        if "decrypting" not in text or "signing" not in text:
            raise Exception("Invalid authority identity file")


        decrypting_key = DecryptingKey.fromPEM(text["decrypting"])
        signing_key = SigningKey.fromPEM(text["signing"])


        return PrivateIdentity(decrypting_key, signing_key)

    def decrypt(self, message):
        ciphertext = CipherText.from_json(message)
        return self.decrypting_key.decrypt(ciphertext)

    def public_identity(self):
        return PublicIdentity(self.decrypting_key.encrypting_key(), self.signing_key.verifying_key())
